### PR TITLE
Adding UnbindStatus to ServiceBindings

### PR DESCRIPTION
**Pull request by n3wscott**
Add a UnbindStatus field to ServiceBindingStatus. The field tracks the status of the ServiceBinding with respect to what needs to be done or has been done for unbinding the ServiceBinding.

When a BindingInstance is created, the UnbindStatus is set to Not Required. This indicates that an unbind request does not need to be made for the ServiceBinding when it is deleted.

When a unbind request is made for a ServiceBinding, the UnbindStatus is set to Required. This indicates that an unbind request does need to be made for the ServiceBinding when it is deleted.

When a unbind request is successful for a ServiceBinding, the UnbindStatus is set to Succeeded. This indicates that no further unbind requests need to be made for the ServiceBinding, and it unblocks the controller from removing the finalizer for the ServiceBinding.

When the controller gives up on sending unbind requests due to failures, the UnbindStatus is set to Failed. This indicates that no further unbind requests need to be made for the ServiceBinding, but it blocks the controller from removing the finalizer for the ServiceBinding. This keeps the resource around so that the user can negotiate unbind directly with the broker. The user must force-delete the ServiceBinding to remove it.

Relates to first point in #1423

### CODE REVIEW

**Comment by n3wscott** [comment]
The unit tests are still being worked on.

**Comment by n3wscott** [comment]
Ready!

File: pkg/controller/controller_binding.go [hunk 1]
```diff
@@ -645,119 +652,234 @@ func (c *controller) reconcileServiceBinding(binding *v1beta1.ServiceBinding) er
...
+			errorWithOngoingAsyncOperationMessage,
+		)
+		if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
+			return false, err
 		}
+		return false, fmt.Errorf("Ongoing Asynchronous operation")
+	}
+
+	if instance.Spec.ClusterServiceClassRef == nil || instance.Spec.ClusterServicePlanRef == nil {
+		// Do not retry if the user wants to cancel/abort a binding request.
```
Comment (staebler) [inline pkg/controller/controller_binding.go]: Looks like this out-of-date comment made its way back in.

I think that ultimately here we need to use logic similar to what is done to determine the plan ID for deprovisioning. We need to allow a ServiceBinding to be deleted, with an unbind request sent to the broker, even if the ServiceInstance has been changed to a non-existent plan. That is outside the scope of this PR, however.

File: pkg/controller/controller_test.go [hunk 2]
```diff
@@ -2486,16 +2494,24 @@ func assertServiceBindingOperationSuccess(t *testing.T, obj runtime.Object, oper
...
 	)
 	switch operation {
 	case v1beta1.ServiceBindingOperationBind:
 		reason = successInjectedBindResultReason
 		readyStatus = v1beta1.ConditionTrue
+		unbindStatus = v1beta1.ServiceBindingUnbindStatusRequired
 	case v1beta1.ServiceBindingOperationUnbind:
 		reason = successUnboundReason
 		readyStatus = v1beta1.ConditionFalse
+		switch originalBinding.Status.UnbindStatus {
```
Comment (staebler) [inline pkg/controller/controller_test.go]: The UnbindStatus should always be succeeded after a successful unbind request.

File: pkg/controller/controller_test.go [hunk 3]
```diff
@@ -2513,15 +2529,21 @@ func assertServiceBindingOperationSuccessWithParameters(t *testing.T, obj runtim
 	case v1beta1.ServiceBindingOperationUnbind:
 		assertServiceBindingExternalPropertiesNil(t, obj)
 	}
+	assertServiceBindingUnbindStatus(t, obj, unbindStatus)
```
Comment (staebler) [inline pkg/controller/controller_test.go]: We need to add assert for the `UnbindStatus` to the other state asserts for ServiceBinding. The goal is for every test to verify that `UnbindStatus` is correct every time that the status of a ServiceBinding is updated.

```
assertServiceBindingErrorBeforeRequest
assertServiceBindingFailedBeforeRequest
assertServiceBindingOperationInProgressWithParameters
assertServiceBindingStartingOrphanMitigation
assertServiceBindingErrorFetchingBinding
assertServiceBindingErrorInjectingCredentials
assertServiceBindingRequestRetriableErrorWithParameters
assertServiceBindingAsyncInProgress
assertServiceBindingOrphanMitigationSuccess
assertServiceBindingOrphanMitigationFailure
```

File: pkg/controller/controller_test.go [hunk 4]
```diff
@@ -2552,6 +2576,7 @@ func assertServiceBindingAsyncBindErrorAfterStateSucceeded(t *testing.T, obj run
 	assertServiceBindingReconciledGeneration(t, obj, originalBinding.Status.ReconciledGeneration)
 	assertServiceBindingExternalPropertiesParameters(t, obj, nil, "")
 	assertServiceBindingInProgressPropertiesNil(t, obj)
+	assertServiceBindingUnbindStatus(t, obj, originalBinding.Status.UnbindStatus)
```
Comment (staebler) [inline pkg/controller/controller_test.go]: "Required"

File: pkg/controller/controller_test.go [hunk 5]
```diff
@@ -2541,6 +2564,7 @@ func assertServiceBindingAsyncBindRetryDurationExceeded(t *testing.T, obj runtim
 	assertServiceBindingReconciledGeneration(t, obj, originalBinding.Status.ReconciledGeneration)
 	assertServiceBindingInProgressPropertiesParameters(t, obj, nil, "")
 	assertServiceBindingExternalPropertiesNil(t, obj)
+	assertServiceBindingUnbindStatus(t, obj, originalBinding.Status.UnbindStatus)
```
Comment (staebler) [inline pkg/controller/controller_test.go]: "Required"

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (n3wscott) [reply pkg/controller/controller_test.go]: that is not true for the case of notrequired.

**Comment by n3wscott** [comment]
@staebler thanks for looking, I am out tomorrow so I will dig back in on monday.

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (staebler) [reply pkg/controller/controller_test.go]: If the UnbindStatus was Not required, then an unbind request should never have been sent.

File: pkg/controller/controller_test.go (hunk 3 above)
Comment (n3wscott) [reply pkg/controller/controller_test.go]: Done.

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (staebler) [reply pkg/controller/controller_test.go]: No, that is not true. This is for the request to the broker. Operation refers to an operation being performed on the broker. That is why operations are Bind and Unbind instead of Create and Delete. That is why current operation gets set right before sending a request to the broker.

File: pkg/controller/controller_binding_test.go [hunk 6]
```diff
@@ -291,6 +294,9 @@ func TestReconcileServiceBindingWithSecretConflict(t *testing.T) {
 			ExternalID:         testServiceBindingGUID,
 			SecretName:         testServiceBindingSecretName,
 		},
+		Status: v1beta1.ServiceBindingStatus{
+			UnbindStatus: v1beta1.ServiceBindingUnbindStatusRequired,
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: This should be NotRequired. The test should be testing a new ServiceBinding for which no broker request has yet been made.

File: pkg/controller/controller_binding_test.go [hunk 7]
```diff
@@ -385,6 +391,9 @@ func TestReconcileServiceBindingWithParameters(t *testing.T) {
...
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: This should be NotRequired. The test should be testing a new ServiceBinding for which no broker request has yet been made.

File: pkg/controller/controller_binding_test.go [hunk 8]
```diff
@@ -591,6 +603,9 @@ func TestReconcileServiceBindingNonbindableClusterServiceClassBindablePlan(t *te
...
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: This should be NotRequired. The test should be testing a new ServiceBinding for which no broker request has yet been made.

File: pkg/controller/controller_binding_test.go [hunk 9]
```diff
@@ -954,22 +982,77 @@ func TestReconcileServiceBindingDelete(t *testing.T) {
...
+			DeletionTimestamp: &metav1.Time{},
+			Finalizers:        []string{v1beta1.FinalizerServiceCatalog},
+			Generation:        1,
+		},
+		Spec: v1beta1.ServiceBindingSpec{
+			ServiceInstanceRef: v1beta1.LocalObjectReference{Name: testServiceInstanceName},
+			ExternalID:         testServiceBindingGUID,
+		},
+		Status: v1beta1.ServiceBindingStatus{
+			UnbindStatus: v1beta1.ServiceBindingUnbindStatusNotRequired,
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: This should be Required. If it is NotRequired, then the reconciler will never get to the point where it checks the service class.

File: pkg/controller/controller_binding_test.go [hunk 10]
```diff
@@ -1207,6 +1291,9 @@ func TestReconcileServiceBindingWithClusterServiceBrokerError(t *testing.T) {
 			ExternalID:         testServiceBindingGUID,
 			SecretName:         testServiceBindingSecretName,
 		},
+		Status: v1beta1.ServiceBindingStatus{
+			UnbindStatus: v1beta1.ServiceBindingUnbindStatusRequired,
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: NotRequired

File: pkg/controller/controller_binding_test.go [hunk 11]
```diff
@@ -1268,6 +1355,9 @@ func TestReconcileServiceBindingWithClusterServiceBrokerHTTPError(t *testing.T)
...
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: NotRequired

File: pkg/controller/controller_binding_test.go [hunk 12]
```diff
@@ -2132,6 +2225,9 @@ func TestReconcileServiceBindingWithSecretParameters(t *testing.T) {
...
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: NotRequired

File: pkg/controller/controller_binding_test.go [hunk 13]
```diff
@@ -2292,82 +2397,87 @@ func TestReconcileBindingWithSetOrphanMitigation(t *testing.T) {
...
+					Namespace:  testNamespace,
+					Generation: 1,
+				},
+				Spec: v1beta1.ServiceBindingSpec{
+					ServiceInstanceRef: v1beta1.LocalObjectReference{Name: testServiceInstanceName},
+					ExternalID:         testServiceBindingGUID,
+					SecretName:         testServiceBindingSecretName,
+				},
+				Status: v1beta1.ServiceBindingStatus{
+					UnbindStatus: v1beta1.ServiceBindingUnbindStatusRequired,
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: NotRequired

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (n3wscott) [reply pkg/controller/controller_test.go]: Oh well I did not know that! Thanks for clearing that up. I have removed that switch statement and the tests still pass so nothing was using that :)

File: pkg/controller/controller_test.go [hunk 14]
```diff
@@ -2470,6 +2481,7 @@ func assertServiceBindingOperationInProgressWithParameters(t *testing.T, obj run
 		assertServiceBindingInProgressPropertiesNil(t, obj)
 	}
 	assertServiceBindingExternalPropertiesUnchanged(t, obj, originalBinding)
+	assertServiceBindingUnbindStatus(t, obj, originalBinding.Status.UnbindStatus)
```
Comment (staebler) [inline pkg/controller/controller_test.go]: The `UnbindStatus` should be Required. If there is an operation is progress, then a Bind request has been sent to the broker and an Unbind request has not yet been successful. Hence, an unbind is required.

File: pkg/controller/controller_test.go [hunk 15]
```diff
@@ -2478,6 +2490,7 @@ func assertServiceBindingStartingOrphanMitigation(t *testing.T, obj runtime.Obje
 	assertServiceBindingOperationStartTimeSet(t, obj, false)
 	assertServiceBindingReconciledGeneration(t, obj, originalBinding.Status.ReconciledGeneration)
 	assertServiceBindingOrphanMitigationSet(t, obj, true)
+	assertServiceBindingUnbindStatus(t, obj, originalBinding.Status.UnbindStatus)
```
Comment (staebler) [inline pkg/controller/controller_test.go]: The `UnbindStatus` should be Required. If an orphan mitigation is in progress, then by definition an unbind is required.

File: pkg/controller/controller_test.go [hunk 16]
```diff
@@ -2649,6 +2685,12 @@ func assertServiceBindingOrphanMitigationSuccess(t *testing.T, obj runtime.Objec
 	assertServiceBindingReconciledGeneration(t, obj, originalBinding.Generation)
 	assertServiceBindingInProgressPropertiesNil(t, obj)
 	assertServiceBindingExternalPropertiesNil(t, obj)
+	switch originalBinding.Status.UnbindStatus {
+	case v1beta1.ServiceBindingUnbindStatusNotRequired:
```
Comment (staebler) [inline pkg/controller/controller_test.go]: The original `UnbindStatus` should never be NotRequired. If orphan mitigation suceeded, then orphan mitigation was in progress for the original binding. Hence, an unbind must have been required for the original binding.

File: pkg/controller/controller_test.go [hunk 17]
```diff
@@ -2659,6 +2701,12 @@ func assertServiceBindingOrphanMitigationFailure(t *testing.T, obj runtime.Objec
...
```
Comment (staebler) [inline pkg/controller/controller_test.go]: As above, the original `UnbindStatus` must be Required.

File: pkg/controller/controller_binding_test.go (hunk 6 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Done.

File: pkg/controller/controller_binding_test.go (hunk 7 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Done.

File: pkg/controller/controller_binding_test.go (hunk 8 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Done.

File: pkg/controller/controller_binding_test.go (hunk 9 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: True! And there was a logic issue in the test expecting an error and also expecting err == nil. weird.

File: pkg/controller/controller_binding_test.go (hunk 10 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Done.

File: pkg/controller/controller_binding_test.go (hunk 11 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Done.

File: pkg/controller/controller_binding_test.go (hunk 12 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: done.

File: pkg/controller/controller_test.go (hunk 14 above)
Comment (n3wscott) [reply pkg/controller/controller_test.go]: Fixed.

File: pkg/controller/controller_binding_test.go (hunk 13 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: Done.

File: pkg/controller/controller_binding_test.go [hunk 18]
```diff
@@ -954,22 +991,77 @@ func TestReconcileServiceBindingDelete(t *testing.T) {
...
+			ExternalID:         testServiceBindingGUID,
+		},
+		Status: v1beta1.ServiceBindingStatus{
+			UnbindStatus: v1beta1.ServiceBindingUnbindStatusRequired,
+		},
+	}
+
+	err := testController.reconcileServiceBinding(binding)
+	if err == nil {
+		t.Fatal("serviceclassref was nil + generation was 1: reconcile should abort the binding")
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: What does "generation was 1" have to do with whether the reconciliation was successful or not? Also, I don't think the binding was aborted, was it? Or maybe I don't know what it means to "abort a binding".

File: pkg/controller/controller_binding_test.go [hunk 19]
```diff
@@ -954,22 +991,77 @@ func TestReconcileServiceBindingDelete(t *testing.T) {
...
+	}
+
+	brokerActions := fakeClusterServiceBrokerClient.Actions()
+	assertNumberOfClusterServiceBrokerActions(t, brokerActions, 0)
+
+	actions := fakeCatalogClient.Actions()
+	// The actions should be:
+	// 0. Clear the finalizer
```
Comment (staebler) [inline pkg/controller/controller_binding_test.go]: Is this true? From looking at the code, it looks like the reconciliation returns an error without doing any status updates to the binding.

File: pkg/controller/controller_binding.go [hunk 20]
```diff
@@ -645,119 +650,234 @@ func (c *controller) reconcileServiceBinding(binding *v1beta1.ServiceBinding) er
...
+			return false, err
 		}
+		return false, fmt.Errorf("Ongoing Asynchronous operation")
+	}
+
+	if instance.Spec.ClusterServiceClassRef == nil || instance.Spec.ClusterServicePlanRef == nil {
+		// TODO(#1562): ultimately here we need to use logic similar to what is done to determine the plan ID for
+		// deprovisioning. We need to allow a ServiceBinding to be deleted, with an unbind request sent to the broker,
+		// even if the ServiceInstance has been changed to a non-existent plan.
+		return false, fmt.Errorf("ClusterServiceClass or ClusterServicePlan references for Instance have not been resolved yet")
```
Comment (staebler) [inline pkg/controller/controller_binding.go]: We should be logging an info message, recording an event, and updating the binding condition before returning here.

File: pkg/controller/controller_binding_test.go (hunk 18 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: This was the test I added to test the case were you bind to a service instance with unresolved plan or class and you then try to delete the binding but can't. 

Binding Unbind status is going back to ServiceBindingUnbindStatusNotRequired, and there is no error expected. It should just delete.

File: pkg/controller/controller_binding_test.go (hunk 19 above)
Comment (n3wscott) [reply pkg/controller/controller_binding_test.go]: For this case it should be just clearing the finalizer with no actions, the unbind status is suppose to be `NotRequired`

File: pkg/controller/controller_binding_test.go (hunk 18 above)
Comment (staebler) [reply pkg/controller/controller_binding_test.go]: Ah, I see. I missed the part where it was testing a binding for which no BindRequest had been sent. My bad.

File: pkg/controller/controller_binding_test.go (hunk 19 above)
Comment (staebler) [reply pkg/controller/controller_binding_test.go]: Yep. I missed the point of this test.

File: pkg/controller/controller_binding.go (hunk 20 above)
Comment (staebler) [reply pkg/controller/controller_binding.go]: This was existing code. This will be dealt with at a later date.

**Comment by kibbles-n-bytes** [comment]
Great work! Thanks for this.
